dali/daliadmin/daliadmin.cpp

```diff
@@ -2794,114 +2798,177 @@ static void wuidCompress(const char *match, const char *type, bool compress)
+
+    std::vector<IPropertyTree *> generateDllsToDelete;
+
+    PROGLOG("Gathering associated files");
+
+    ts.reset(0);
+    Owned<IRemoteConnection> generatedDllsConn = querySDS().connect("/GeneratedDlls", myProcessSession(), fix?RTM_LOCK_WRITE:RTM_LOCK_READ, 10000);
+    if (!generatedDllsConn)
+        throw makeStringException(0, "Failed to connect to /GeneratedDlls");
+    IPropertyTree *root = generatedDllsConn->queryRoot()->queryBranch(NULL); // force everything to be fetched
 
     Owned<IPropertyTreeIterator> gdIter = root->getElements("*");
     RegExpr RE("^.*{W2[0-9][0-9][0-9][0-9][0-9][0-9][0-9]-[0-9][0-9][0-9][0-9][0-9][0-9]{-[0-9]+}?}{[^0-9].*|}$");
 
-    unsigned index=1;
+    unsigned removedLocations = 0;
+    unsigned totalCount = 0;
     ForEach(*gdIter)
     {
+        ++totalCount;
         IPropertyTree &gd = gdIter->query();
-        const char *name = gd.queryProp("@name");
-        if (name && *name)
+        const char *eName = gd.queryName();
+        const char *name = nullptr;
+        if (streq(eName, "GeneratedDll")) // legacy format
+        {
+            name = gd.queryProp("@name");
+            if (isEmptyString(name)) // would mean invalid entry if so
+                continue;
+        }
+        else if (startsWith(eName, "Dll")) // uniq entry name
+        {
+            name = gd.queryProp("@name");
+            if (isEmptyString(name)) // would mean invalid entry if so
+                continue;
+        }
+        else // unknown
+        {
+            continue;
+        }
+
+        bool associatedWithWuid = false;
+        bool entryRemoved = false;
+        if (RE.find(name)) // if it's a workunit associated dll/file
         {
-            if (RE.find(name))
+            StringBuffer wuid;
+            RE.substitute(wuid,"#1");
+            const char *w = wuid.str();
+            associatedWithWuid = (nullptr != wuids.find(*w));
+            if (!associatedWithWuid)
             {
-                StringBuffer wuid;
-                RE.substitute(wuid,"#1");
-                const char *w = wuid.str();
-                bool found = NULL != wuids.find(*w);
-                const char *uid = gd.queryProp("@uid");
-                if (!found)
+                generateDllsToDelete.push_back(&gd);
+                entryRemoved = true;
+            }
+        }
+
+        if (!entryRemoved && removeStaleLocations)
+        {
+            std::vector<IPropertyTree *> locationsToRemove;
+            Owned<IPropertyTreeIterator> locationIter = gd.getElements("location");
+            ForEach(*locationIter)
+            {
+                IPropertyTree &location = locationIter->query();
+                const char *host = location.queryProp("@ip");
+                if (!isEmptyString(host))
                 {
-                    uidsToDelete.append(uid);
-                    indexToDelete.append(index);
+                    IpAddress ip(host);
```

Comment:
what would happen if hosts were stored by name, and that host no longer existed?

Reply:
it would continue to work afaics.
If hosts in machine map and here, they will be bound at the point it builds the map (hosts->ip), and bound here, so this lookup would succeed whether hosts or IPs.